### PR TITLE
AP_HAL_ChibiOS: Add Diatone MambaF405 MK2

### PR DESCRIPTION
**Pull request by GHswitt**
This is based on the Mamba F405 PR https://github.com/ArduPilot/ardupilot/pull/15392
Board available from: https://www.diatone.us/products/mamba-f405-flight-controller-mk2

I have tested the following with a copter:

- IMU
- SBUS input
- RC out 1-4 (in DShot150 mode)
- Battery voltage and current
- Dataflash
- Serial 3 (GPS)
- Serial 6 (Frsky Passthrough)
- I2C (external compass and barometer)
- OSD

Please check if anything doesn't fit.

Thanks
Sebastian

### CODE REVIEW

**Comment by GHswitt** [comment]
So, is something missing?

File: libraries/AP_HAL_ChibiOS/hwdef/MambaF405v2/hwdef.dat [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+define HAL_OSD_TYPE_DEFAULT 1
+# Font for the osd
+ROMFS_WILDCARD libraries/AP_OSD/fonts/font*.bin
+
+# To complementary channels work we define this
+#define STM32_PWM_USE_ADVANCED TRUE
+
+define BOARD_PWM_COUNT_DEFAULT 4
+
+#define CH_DBG_ENABLE_ASSERTS TRUE
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/MambaF405v2/hwdef.dat:160]: You don't want these

**Review by andyp1per — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/MambaF405v2/hwdef.dat (hunk 1 above)
Comment (GHswitt) [reply libraries/AP_HAL_ChibiOS/hwdef/MambaF405v2/hwdef.dat:160]: Done

File: libraries/AP_HAL_ChibiOS/hwdef/MambaF405v2/hwdef.dat [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/MambaF405v2/hwdef.dat:160]: You don't want this either. Users can enable asserts via --enable-asserts to waf

Comment (GHswitt) [reply libraries/AP_HAL_ChibiOS/hwdef/MambaF405v2/hwdef.dat]: Ok, thank you.

**Comment by andyp1per** [comment]
Needs a squash into a single commit and a rebase. The commit message needs to start with the module name, so "AP_HAL_ChibiOS:"

**Comment by GHswitt** [comment]
Done

**Comment by andyp1per** [comment]
... and of course I have given you a slightly bum steer here - sorry : ( The tools changes should be in a separate commit labeled "Tools: ..."

**Comment by GHswitt** [comment]
And now I learned how to split a commit ;)

**Review by andyp1per — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
@GHswitt thanks for a great contribution!
@Hwurzburg may be in touch about adding this to our wiki

**Comment by GHswitt** [comment]
@tridge @andyp1per Thank you and the whole team for your great work!
No problem, just leave a message.

**Comment by lavachemist** [comment]
@GHswitt are you still having success with this? 

I got the Mamba F405 MK2 working partially on a wing, but I kept having issues with I2C, which I was able to resolve by commenting out the line for the BMP280 (I don't understand the BMP280 being in there since this board does not have an onboard baro). Then I started having GPS issues once the baro was working (possibly coincidental timing), then it crashed completely and now it won't connect to Mission Planner.

Also, using the PPM pad as RX1 for non-inverted UART did not seem to work, so I had to use UART 6 for RC link. Any ideas on that one? I tried both alternate board configs and neither seemed to work.

**Comment by GHswitt** [comment]
Yes, I'm flying it on a copter with GPS and BMP280 baro. The section:
`# Probe for I2C BMP280, but allow init on board variants without onboard baro too
BARO BMP280 I2C:0:0x76
define HAL_PROBE_EXTERNAL_I2C_BAROS
define HAL_BARO_ALLOW_INIT_NO_BARO
`
was copied from another hwdef.dat. I thought that HAL_BARO_ALLOW_INIT_NO_BARO should also allow it working without baro? 

I'm using SBUS and Frsky telemetry on SERIAL6. 

Have you tried using SERIAL1_OPTIONS=1 (InvertRX)? That's what `PC0 USART1_RXINV OUTPUT HIGH GPIO(78) POL(0)` should do.

**Comment by lavachemist** [comment]
> Yes, I'm flying it on a copter with GPS and BMP280 baro. The section:
> `# Probe for I2C BMP280, but allow init on board variants without onboard baro too BARO BMP280 I2C:0:0x76 define HAL_PROBE_EXTERNAL_I2C_BAROS define HAL_BARO_ALLOW_INIT_NO_BARO `
> was copied from another hwdef.dat. I thought that HAL_BARO_ALLOW_INIT_NO_BARO should also allow it working without baro?
> 
> I'm using SBUS and Frsky telemetry on SERIAL6.
> 
> Have you tried using SERIAL1_OPTIONS=1 (InvertRX)? That's what `PC0 USART1_RXINV OUTPUT HIGH GPIO(78) POL(0)` should do.

I will try the baro again, flashing from scratch, but I am using a different baro model (MS5611) so that might have been why yours worked and mine did not. 

I will try serial inversion command, I wouldn't have intuitively expected that to work since the F405 doesn't support inverted protocols. That would be awesome if it fixes UART 1. Thanks for the reply!

**Comment by GHswitt** [comment]
> I will try serial inversion command, I wouldn't have intuitively expected that to work since the F405 doesn't support inverted protocols. That would be awesome if it fixes UART 1. Thanks for the reply!

There is a hardware circuit for this one input signal, other pins do not work.

**Comment by Hwurzburg** [comment]
@GHswitt I am starting to work on adding this to the wiki....but I have a question...have you tried using the PPM input for RC yet instead of the SBUS input? since this board apparently has a controlled (implying pushpull output) inverter between SBUS and RX1 (PPM pin), instead of an open drain fixed inverter  like other boards (which will not interfere with signals on the RX1 pin if SBUS is left floating)...I am concerned that PPM will not work as an RC input (or as a pure RX1 in ALT1) unless the inverter is intentionally forced off...

**Comment by GHswitt** [comment]
@Hwurzburg You are right, I think it's PPM or SBUS, depending on PC0. I've not used PPM, I can try next weekend.

**Comment by lavachemist** [comment]
@GHswitt I haven't had a chance to try the "SERIAL1_OPTIONS=1" command yet, but I did get the Mamba F405 MK2 working on a wing today. 

The Baro configuration appears to be incorrect because Ardu does not find an MS5611 on the external bus. I am able to get Ardu to recognize it by manually defining the baro_ext_bus and baro_probe_ext parameters.

With those changes, I have it working properly, although I have not maidened the wing yet.

**Comment by lavachemist** [comment]
@GHswitt have you had any luck getting univerted RX on UART 1? I can't seem to figure it out. "SERIAL1_OPTIONS=1" does not seem to do anything as far as I can tell. I've tried both the PPM and SBUS pad.

**Comment by lavachemist** [comment]
I was not able to find a command that would make serial 1 RX uninverted. I know there is a way because I can edit the target like this, and force it to work:

`# USART1 is wired to SBUS and PPM input
# RC input defaults to timer capture
# PA10 TIM1_CH3 TIM1 RCININT PULLDOWN
PA9 USART1_TX USART1 NODMA

# Alt config to allow RCIN on UART
PA10 USART1_RX USART1 NODMA

# SBUS inversion control pin, active low
PC0 USART1_RXINV OUTPUT LOW GPIO(78) POL(0)`

**Comment by Sekilsgs2** [comment]
> > Yes, I'm flying it on a copter with GPS and BMP280 baro. The section:
> > `# Probe for I2C BMP280, but allow init on board variants without onboard baro too BARO BMP280 I2C:0:0x76 define HAL_PROBE_EXTERNAL_I2C_BAROS define HAL_BARO_ALLOW_INIT_NO_BARO `
> > was copied from another hwdef.dat. I thought that HAL_BARO_ALLOW_INIT_NO_BARO should also allow it working without baro?
> > I'm using SBUS and Frsky telemetry on SERIAL6.
> > Have you tried using SERIAL1_OPTIONS=1 (InvertRX)? That's what `PC0 USART1_RXINV OUTPUT HIGH GPIO(78) POL(0)` should do.
> 
> I will try the baro again, flashing from scratch, but I am using a different baro model (MS5611) so that might have been why yours worked and mine did not.
> 
> I will try serial inversion command, I wouldn't have intuitively expected that to work since the F405 doesn't support inverted protocols. That would be awesome if it fixes UART 1. Thanks for the reply!

Hi. please help me - how you run ms5611 - what im need change? and my gps on serial6 not working - in inav working good...

**Comment by GHswitt** [comment]
@lavachemist Sorry, I did not get it to work via SERIAL1_OPTIONS either. It seems it does not have an influence on PC0 USART1_RXINV. I'm not so far in to know why...maybe a dev could answer that?

@Sekilsgs2 Have you set SERIAL6_PROTOCOL to 5 (GPS)? I'm using 10 (Frsky Passthrough/OpenTX) with it so I know the port works. Also check if it is required to change the SERIAL6_BAUD.
As for the ms5611 (isn't that a SPI device?) I think one needs to change the hwdev. I've only added BMP280 support and added HAL_PROBE_EXTERNAL_I2C_BAROS (which I thought checks also other baros, but I may be wrong).

# Probe for I2C BMP280, but allow init on board variants without onboard baro too
BARO BMP280 I2C:0:0x76
define HAL_PROBE_EXTERNAL_I2C_BAROS
define HAL_BARO_ALLOW_INIT_NO_BARO

**Comment by lavachemist** [comment]
@GHswitt I hope we can figure it out, because getting rid of the inversion made it possible for me to run CRSF. I wouldn't have been able to use it otherwise, since there are only three serial interfaces. I currently have this FC in a fixed wing and it works great. I just wish that I didn't have to edit the target to make it work for me.

**Comment by Sekilsgs2** [comment]
> @Sekilsgs2 Have you set SERIAL6_PROTOCOL to 5 (GPS)? I'm using 10 (Frsky Passthrough/OpenTX) with it so I know the port works. Also check if it is required to change the SERIAL6_BAUD.
> As for the ms5611 (isn't that a SPI device?) I think one needs to change the hwdev. I've only added BMP280 support and added HAL_PROBE_EXTERNAL_I2C_BAROS (which I thought checks also other baros, but I may be wrong).

Im fix issue - need disable GPS an all other (fake ports) - i'm think need make fake ports with state -1.
ms5611 i2c and cant detected.  Only if i'm change external bus and id for ms5611.
And i'm think bmp280 no need in hwdef - because this board dont have any baros..
And i'm think need change motors channel order to make ardupilot compatible to inav, betaflight - but now i'm need change order for each motor because this not compatible...

**Pull request merged**